src/python/pants/engine/target.py

```diff
@@ -868,6 +848,27 @@ def __init__(
         self.expect_single_field_set = expect_single_field_set
 
 
+@frozen_after_init
+@dataclass(unsafe_hash=True)
+class FieldSetsPerTarget(Generic[_AFS]):
+    # One tuple of FieldSet instances per input target.
+    collection: Tuple[Tuple[_AFS, ...], ...]
+
+    def __init__(self, collection: Iterable[Iterable[_AFS]]):
+        self.collection = tuple(tuple(iterable) for iterable in collection)
+
```

Comment:
The properties `field_sets` and `targets` have been useful. You may want to add them here.

Reply:
The targets aren't known here. And the `collection` field is just the fieldsets. Note that there's no mapping here.

Reply:
Oh. What's the reason to get rid of the mapping and the targets?

Reply:
This factored-out piece of functionality doesn't need them.